### PR TITLE
FEAT: MultiGPU for golang bindings

### PR DESCRIPTION
**Pull request by jeremyfelder**
## Describe the changes

This PR adds multi gpu support in the golang bindings.

Tha main changes are to DeviceSlice which now includes a `deviceId` attribute specifying which device the underlying data resides on and checks for correct deviceId and current device when using DeviceSlices in any operation.

In Go, most concurrency can be done via Goroutines (described as lightweight threads - in reality, more of a threadpool manager), however, there is no guarantee that a goroutine stays on a specific host thread. Therefore, a function `RunOnDevice` was added to the cuda_runtime package which locks a goroutine into a specific host thread, sets a current GPU device, runs a provided function, and unlocks the goroutine from the host thread after the provided function finishes. While the goroutine is locked to the hsot thread, the Go runtime will not assign other goroutines to that host thread

### CODE REVIEW

**Comment by DmytroTym** [comment]
@jeremyfelder one thing I found really useful when designing Rust-side device slices is the ability to look up device id of any device pointer using CUDA runtime [cudaPointerGetAttributes](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED_1gd89830e17d399c064a2f3c3fa8bb4390) function. This allows to avoid explicitly storing device id alongside the actual data.

File: wrappers/golang/core/slice.go [hunk 1]
```diff
@@ -43,8 +43,16 @@ func (d DeviceSlice) IsOnDevice() bool {
 	return true
 }
 
-// TODO: change signature to be Malloc(element, numElements)
-// calc size internally
+func (d DeviceSlice) GetDeviceId() int {
+	return cr.GetDeviceFromPointer(d.inner)
+}
+
+func (d DeviceSlice) CheckDevice() {
```
Comment (ImmanuelSegol) [inline wrappers/golang/core/slice.go:50]: document this with a comment please

File: wrappers/golang/cuda_runtime/device_context.go [hunk 2]
```diff
@@ -17,20 +18,24 @@ type DeviceContext struct {
...
 
 	/// Mempool to use. Default value: 0.
-	// TODO: use cuda_bindings.CudaMemPool as type
-	Mempool uint // Assuming the type is provided by a CUDA binding crate
+	Mempool MemPool // Assuming the type is provided by a CUDA binding crate
 }
 
 func GetDefaultDeviceContext() (DeviceContext, CudaError) {
+	device, err := GetDevice()
+	if err != CudaSuccess {
```
Comment (ImmanuelSegol) [inline wrappers/golang/cuda_runtime/device_context.go:29]: is `err` a descriptive error?

if so perhaps its worth adding it to the panic.

**Review by ImmanuelSegol — CHANGES_REQUESTED** [review]

File: wrappers/golang/internal/generator/templates/curve.go.tmpl [hunk 3]
```diff
@@ -169,10 +171,12 @@ func convert{{if .IsG2}}G2{{end}}ProjectivePointsMontgomery(points *core.DeviceS
...
 func {{if .IsG2}}G2{{end}}ProjectiveToMontgomery(points *core.DeviceSlice) cr.CudaError {
+	points.CheckDevice()
 	return convert{{if .IsG2}}G2{{end}}ProjectivePointsMontgomery(points, true)
 }
 
 func {{if .IsG2}}G2{{end}}ProjectiveFromMontgomery(points *core.DeviceSlice) cr.CudaError {
+	points.CheckDevice()
 	return convert{{if .IsG2}}G2{{end}}ProjectivePointsMontgomery(points, false)
 }
 {{end}}
```
Comment (ImmanuelSegol) [inline wrappers/golang/internal/generator/templates/curve.go.tmpl:182]: missing extra line

File: wrappers/golang/internal/generator/templates/scalar_field.go.tmpl [hunk 4]
```diff
@@ -33,9 +33,11 @@ func convertScalarsMontgomery(scalars *core.DeviceSlice, isInto bool) cr.CudaErr
...
 
 func ToMontgomery(scalars *core.DeviceSlice) cr.CudaError {
+	scalars.CheckDevice()
 	return convertScalarsMontgomery(scalars, true)
 }
 
 func FromMontgomery(scalars *core.DeviceSlice) cr.CudaError {
+	scalars.CheckDevice()
 	return convertScalarsMontgomery(scalars, false)
 }{{- end}}
```
Comment (ImmanuelSegol) [inline wrappers/golang/internal/generator/templates/scalar_field.go.tmpl:43]: extra line

**Review by ImmanuelSegol — APPROVED** [review]
its the same design as the rust version of multigpu so no comments on design.

The code looks good. no blocking comments so i just added some suggestions.

**Review by LeonHibnik — APPROVED** [review]

**Pull request merged**